### ISSUE TITLE
Make Samplers have own Random number generator, not use the global one

### ISSUE THREAD

**Issue by oxinabox** [baggepinnen/Hyperopt.jl#32]
Right now the samplers call out to `rand` which uses the global random number generator.

However, lots of models themselves use the global random number generator  (e.g. initial NN weights).
To work around that the nondetermanist its common call `Random.seed!(1)` at the start of each model evaluation.

However, this breaks HyperOpt.jl it it will always have the same seed when it is called.
This is particular apparent in random search, which will not search, just keep retrying same parameters.

Solution to this would be to give reach Sampler its own random number generator.
Which would be initialized with a seed (default to `rand(UInt)`?

Then calling `Random.seed!(1)` would not change anything in the behavour of the hyperoptimizer

**Comment by baggepinnen** [comment]
That sounds like a good idea! In the meantime you could perhaps make use of the `LHsampler` since this does all it's work choosing samples before it ever calls the function.

**Comment by luboshanus** [comment]
Hi, 
is there any chance RandomSampler will work when using Random.seed!()?  Or should I turn of Random.seed!(xxxx) ? I don't like the LHSampler() since I don't have the same length of vectors of parameters over which I search optimum. 
Thanks a lot!

**Comment by baggepinnen** [comment]
> is there any chance RandomSampler will work when using Random.seed!()?

Yes, it's for sure possible to solve this, like Lyndon suggested above. I'm not likely to find the time to do so any time soon, but feel free to submit a PR and I'll happily merge it into Hyperopt :)

**Comment by baggepinnen** [comment]
Hyperopt now has an internal, thread local `HO_RNG` that is used for randomness inside the package. This should hopefully resolve the issues reported here :)

**Comment by luboshanus** [comment]
Hi, thanks for the message. It is great it works now!

I have an additional question related to `@phyperopt` and the random searching over the grid of parameters. Below, I provide the candidates run by `@phyperopt` at 50 cores. I know is a bit off-topic here but I run the test you have and the parallel opt works meaning that it searcher random candidates correctly! However, it tests it on 2 cores... 

Thanks!

```
julia> hpo.history
50-element Array{Any,1}:
 (0.001f0, 0.0001f0, 0.5f0)
 (0.001f0, 0.0001f0, 0.5f0)
 (0.001f0, 0.0001f0, 0.5f0)
 ⋮
 (0.001f0, 0.0001f0, 0.5f0)
 (0.001f0, 0.0001f0, 0.5f0)
```

**Comment by luboshanus** [comment]
It might be related to #38 .

I am obtaining log as below:
```
> Parallel Hyperopt started:  2021-01-07T01:23:26.04
      From worker 11:	  ★ Candidate 8 	Pars: η=0.005 λ=0.0001  ϕ=0.3
      From worker 7:      ★ Candidate 6 	Pars: η=0.005 λ=0.0001  ϕ=0.3
      From worker 2:	  ★ Candidate 7 	Pars: η=0.005 λ=0.0001  ϕ=0.3
      From worker 9:	  ★ Candidate 1 	Pars: η=0.005 λ=0.0001  ϕ=0.3
      From worker 5:	  ★ Candidate 2 	Pars: η=0.005 λ=0.0001  ϕ=0.3
      From worker 4:	  ★ Candidate 4 	Pars: η=0.005 λ=0.0001  ϕ=0.3
      From worker 3:	  ★ Candidate 9 	Pars: η=0.005 λ=0.0001  ϕ=0.3
      From worker 10:	  ★ Candidate 3 	Pars: η=0.005 λ=0.0001  ϕ=0.3
```
When adding 10 processes, and if I want 20 iterations to be done by `@phyperopt`, then the first 10 is like this and the second 10 is a new set of params but at all worker same again. Looks like a strage sampled set in `ho.history`.

The number following Candidate _  is `i` from the loop in `@phyperopt`.

**Comment by luboshanus** [comment]
I have tested `LHSampler()` and it takes different (unique) sets of parameters but `RandomSamplere()` does not. So, I checked how both are defined in `src/samplers.jl`

RandomSampler():
```
function init!(::RandomSampler, ho) end

function (s::RandomSampler)(ho, iter)
    [list[rand(HO_RNG[threadid()], 1:length(list))] for list in ho.candidates]
end
```

LHSampler():
```
function init!(s::LHSampler, ho)
    # code doing LHSampler...

end

function (s::LHSampler)(ho, iter)
    init!(s, ho)
    [list[s.samples[dim,iter]] for (dim,list) in enumerate(ho.candidates)]
end
```

I don't understand those structures properly, but the first thing I see is that function `init!()` misses `s` before  `::RandomSampler`, and while running the `threadid()` is 1 for all workers. But still does not understand why `@phyperopt` does not properly initialize candidates. 

Thanks for checking. Also, if these observations are irrelevant, my apologies.

**Comment by ericphanson** [comment]
I think this is closed by #58
